### PR TITLE
Monit 26426 delta histogram

### PR DESCRIPTION
**Pull request by deosu**
Proxy understands OTLP Delta Histograms and transforms them to WF specific type.

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 1]
```diff
@@ -174,24 +184,99 @@ private static List<ReportPoint> transformHistogram(String name, Histogram histo
...
+      return point.getExplicitBounds(currentIndex - 1);
+    }
+    return (point.getExplicitBounds(currentIndex - 1) +
+        point.getExplicitBounds(currentIndex)) / 2.0;
+  }
+
+  private static List<ReportPoint> transformCumulativeHistogramDataPoint(String name,
+                                                                         HistogramDataPoint point, List<KeyValue> resourceAttrs) {
     if (point.getExplicitBoundsCount() != point.getBucketCountsCount() - 1) {
       throw new IllegalArgumentException("OTel: histogram " + name + " data point has mismatched " +
```
Comment (keep94) [inline proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:281]: I know this code isn't part of your PR, but this message is confusing to me too for the same reason. Maybe fix it like the one above.

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java [hunk 2]
```diff
@@ -342,17 +351,97 @@ public void cumulativeHistogram() {
         .setAnnotations(Collections.singletonMap("le", "+Inf"))
         .build();
 
-    mockMetricsHandler.report(wfMetric1);
-    mockMetricsHandler.report(wfMetric2);
-    mockMetricsHandler.report(wfMetric3);
+    mockReportPointHandler.report(wfMetric1);
+    mockReportPointHandler.report(wfMetric2);
+    mockReportPointHandler.report(wfMetric3);
+    EasyMock.expectLastCall();
```
Comment (keep94) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java:357]: For my own education, why do we need expectLastCall here?  It looks like there are three calls expected.  Why don't we have to call expectLastCall three times?

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java [hunk 3]
```diff
@@ -342,17 +351,97 @@ public void cumulativeHistogram() {
...
+        .setTimestamp(TimeUnit.NANOSECONDS.toMillis(epochTime))
+        .setValue(hourHistogram)
+        .setHost(DEFAULT_SOURCE)
+        .build();
+
+    wavefront.report.Histogram dayHistogram = wavefront.report.Histogram.newBuilder().
+        setType(HistogramType.TDIGEST).
+        setBins(bins).
+        setCounts(counts).
+        setDuration(86400000).
```
Comment (keep94) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java:424]: Consider writing as 24\*60\*60\*1000. and consider assigning to a constant such as MILLIS_IN_DAY.  Do the same for the others.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 4]
```diff
@@ -174,24 +184,99 @@ private static List<ReportPoint> transformHistogram(String name, Histogram histo
...
+      int duration;
+      switch (granularity) {
+        case MINUTE:
+          duration = 60000;
+          break;
+        case HOUR:
+          duration = 3600000;
+          break;
+        case DAY:
+          duration = 86400000;
```
Comment (keep94) [inline proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:242]: Consider writing as 24\*60\*60\*1000 and assigning to a constant such as MILLIS_IN_DAY.  Do the same for the others and have MILLIS_IN_HOUR and MILLIS_IN_MINUTE.

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java [hunk 5]
```diff
@@ -427,6 +429,125 @@ public void transformsCumulativeHistogramThrowsMalformedDataPointsError() {
             emptyAttrs, null, DEFAULT_SOURCE));
   }
 
+  @Test
+  public void transformsMinimalDeltaHistogram() {
```
Comment (keep94) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:433]: Just curious why we need this test.  It seems very similar to the one just above it.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 6]
```diff
@@ -53,8 +57,14 @@ public static void exportToWavefront(ExportMetricsServiceRequest request,
 
     for (ReportPoint point : fromOtlpRequest(request, preprocessor, defaultSource)) {
       // TODO: handle sampler
-      if (!wasFilteredByPreprocessor(point, handler, preprocessor)) {
-        handler.report(point);
+      if (point.getValue() instanceof wavefront.report.Histogram) {
+        if (!wasFilteredByPreprocessor(point, histogramHandler, preprocessor)) {
+          histogramHandler.report(point);
+        }
+      } else {
```
[the remark this answers is not in the document]
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:64]: That is a good point. ReportPoint's value is of type Object. For delta histogram we usually get wavefront.Histogram as a value. For other cases, it is a numerical value. But I will make changes.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 7]
```diff
@@ -174,24 +184,99 @@ private static List<ReportPoint> transformHistogram(String name, Histogram histo
...
+      reportPoints.addAll(transformCumulativeHistogramDataPoint(name, point, resourceAttrs));
+    }
+
+    return reportPoints;
+  }
+
+  private static List<ReportPoint> transformDeltaHistogramDataPoint(String name,
+                                                                    HistogramDataPoint point, List<KeyValue> resourceAttrs) {
+    if (point.getExplicitBoundsCount() != point.getBucketCountsCount() - 1) {
+      throw new IllegalArgumentException("OTel: histogram " + name + " data point has mismatched " +
```
[the remark this answers is not in the document]
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:217]: Sounds good, making changes.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 8]
```diff
@@ -174,24 +184,99 @@ private static List<ReportPoint> transformHistogram(String name, Histogram histo
...
+          "explicit bound counts (" + point.getExplicitBoundsCount() + ") and bucket counts (" +
+          (point.getBucketCountsCount() - 1) + ")");
+    }
+
+    List<ReportPoint> reportPoints = new ArrayList<>();
+
+    List<Double> bins = new ArrayList<>();
+    List<Integer> counts = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:225]: Done

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java (hunk 1 above)
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:281]: Done

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java (hunk 3 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java:424]: Done

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java (hunk 4 above)
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:242]: Done

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java (hunk 5 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:433]: Which one? The one above is transformsCumulativeHistogramThrowsMalformedDataPointsError().

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java (hunk 2 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java]: Good catch. I have added EasyMock.expectLastCall().once();

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java (hunk 5 above)
Comment (keep94) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:436]: This seems similar to the deltaHistogram() test.

**Review by keep94 — APPROVED** [review]

**Pull request merged**